File: arrays-and-hashing/217/solve.ts
// npx tsx arrays-and-hashing/217/solve.ts

// Runtime: 20ms, Memory: 75.4MB
function containsDuplicate(nums: number[]): boolean {
  let values = new Set();

  for (const value of nums) {
    if (values.has(value)) {
      console.log(true);
      return true;
    }
    values.add(value);
  }

  console.log(false);
  return false;
}

// Runtime: 17ms, Memory: 74.4MB
function containsDuplicate2(nums: number[]): boolean {
  const values = new Set();

  for (let value of nums) {
    if (values.has(value)) {
      console.log(true);
      return true;
    }
    values.add(value);
  }
  console.log(false);
  return false;
}

// Runtime: 9ms, Memory: 75MB
function containsDuplicate3(nums: number[]): boolean {
  const values = new Set<number>();

  for (let value of nums) {
    if (values.has(value)) {
      console.log(true);
      return true;
    }
    values.add(value);
  }
  console.log(false);
  return false;
}

// Runtime: 18ms, Memory: 74.6MB
function containsDuplicate4(nums: number[]): boolean {
  console.log(new Set(nums).size !== nums.length);
  return new Set(nums).size !== nums.length;
}

console.log("Test 1");
containsDuplicate4([1, 2, 3, 1]);

console.log("Test 2");
containsDuplicate4([1, 2, 3, 4]);

console.log("Test 3");
containsDuplicate4([1, 1, 1, 3, 3, 4, 3, 2, 4, 2]);

type ContainsDuplicate<T extends readonly unknown[]> = T extends readonly [
  infer First,
  ...infer Rest
]
  ? First extends Rest[number]
    ? true
    : ContainsDuplicate<Rest>
  : false;

type Test1 = ContainsDuplicate<[1, 2, 3, 1]>; // true
type Test2 = ContainsDuplicate<[1, 2, 3, 4]>; // false
type Test3 = ContainsDuplicate<[1, 1, 1, 3, 3, 4, 3, 2, 4, 2]>; // true

File: problems/217/solve.go
// go run problems/217/solve.go

package main

import "fmt"

// Runtime: 17ms, Memory: 10.9MB
func containsDuplicate(nums []int) bool {
	values := make(map[int]bool)
	for _, value := range nums {
		if _, ok := values[value]; ok {
			fmt.Println(true)
			return true
		}
		values[value] = true
	}
	fmt.Println(false)
	return false
}

// Runtime: 24ms, Memory: 19.1MB
func containsDuplicate2(nums []int) bool {
	values := make(map[int]struct{})
	for _, value := range nums {
		if _, ok := values[value]; ok {
			fmt.Println(true)
			return true
		}
		values[value] = struct{}{}
	}
	fmt.Println(false)
	return false
}

func main() {
	fmt.Println("Test 1")
	containsDuplicate2([]int{1, 2, 3, 1})
	fmt.Println("Test 2")
	containsDuplicate2([]int{1, 2, 3, 4})
	fmt.Println("Test 3")
	containsDuplicate2([]int{1, 1, 1, 3, 3, 4, 3, 2, 4, 2})
}
